icu4c/source/i18n/unitsdata.cpp

```diff
@@ -94,8 +95,247 @@ class ConversionRateDataSink : public ResourceSink {
+                            } else if (uprv_strcmp(key, "skeleton") == 0) {
+                                int32_t length;
+                                const UChar *s = value.getString(length, status);
+                                up->skeleton.appendInvariantChars(s, length, status);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+  private:
+    MaybeStackVector<UnitPreference> *preferences;
+    MaybeStackVector<UnitPreferenceMetadata> *metadata;
+};
+
+int32_t binarySearch(const MaybeStackVector<UnitPreferenceMetadata> *metadata, const char *category,
+                     const char *usage, const char *region, bool *foundCategory, bool *foundUsage,
+                     bool *foundRegion, UErrorCode &status) {
+    UnitPreferenceMetadata desired(category, usage, region, -1, -1, status);
+    if (U_FAILURE(status)) { return -1; }
+    int32_t start = 0;
+    int32_t end = metadata->length();
+    *foundCategory = false;
+    *foundUsage = false;
+    *foundRegion = false;
+    while (start < end) {
+        int32_t mid = (start + end) / 2;
+        int32_t cmp = compareUnitPreferenceMetadata(*(*metadata)[mid], desired, foundCategory,
+                                                    foundUsage, foundRegion);
+        if (cmp < 0) {
+            start = mid + 1;
+        } else if (cmp > 0) {
+            end = mid;
+        } else {
+            return mid;
+        }
+    }
+    return -1;
+}
+
+/**
+ * Finds the UnitPreferenceMetadata instance that matches the given category,
+ * usage and region: if missing, region falls back to "001", and usage falls
+ * back to "default".
+ *
+ * This is implemented as a binary search, with fallback restarting the search
+ * from the search range at which the parent in the category/usage/region
+ * hierarchy was found.
+ *
+ * @param metadata The full list of UnitPreferenceMetadata instances.
+ * @param category The category to search for. If category is not known, it can
+ * be resolved from the baseunit of the input (for supported unit categories).
+ * TODO(hugovdm): implement the unit->category lookup (via "unitQuantities" in
+ * the units resource bundle).
+ * @param usage The usage for which formatting preferences is needed. If the
+ * given usage is not known, this function automatically falls back to "default"
+ * usage.
+ * @param region The region for which preferences are needed. If there are no
+ * region-specific preferences, this function automatically falls back to the
+ * "001" region (global).
+ * @param status The standard ICU error code output parameter. If an invalid
+ * category is given, status will be U_ILLEGAL_ARGUMENT_ERROR. If fallback to
+ * "default" or "001" didn't resolve, status will be U_MISSING_RESOURCE.
+ * @return The index into the metadata vector which represents the appropriate
+ * preferences. If appropriate preferences are not found, -1 is returned.
```

Comment:
Your comments are great and useful. I prefer to put the illustration in points, instead of a big paragraph.

For example:

```
 * @param status The standard ICU error code output parameter. 
 *     - If an invalid category is given, the status will be U_ILLEGAL_ARGUMENT_ERROR. 
 *     - If fallback to "default" or "001" didn't resolve, the status will be U_MISSING_RESOURCE.
```

Reply:
Done.

Can't test the markdown syntax with "make doc", I think doxygen won't generate output for internal APIs. But that also means syntax isn't going to matter as much either. :-) I'll check "make doc" output for public API docs though.

Reply:
Looks like this is resolved.  Going to mark it as such.